pennylane/ops/qubit.py

```diff
@@ -2157,6 +2157,43 @@ def _matrix(cls, *params):
 
         return U
 
+    @staticmethod
+    def decomposition(U, wires):
+        # Decompose arbitrary single-qubit unitaries as the form RZ RY RZ
+        if U.shape[0] == 2:
+            wires = Wires(wires)
+
+            # Remove the global phase if present; cannot just divide by square root of det
+            # because sometimes it is negative.
+            det = U[0, 0] * U[1, 1] - U[0, 1] * U[1, 0]
+            if not qml.math.isclose(det, 1):
+                U = U * (qml.math.exp(-1j * qml.math.angle(det) / 2))
+
+            # Compute the angle of the RY
+            theta = 2 * qml.math.arccos(qml.math.sqrt(U[0, 0] * U[1, 1]))
+
+            # If it's close to 0, the matrix is diagonal and we have just an RZ rotation
+            if qml.math.isclose(theta, 0, atol=1e-7):
+                omega = 2 * qml.math.angle(U[1, 1])
+                return [RZ(omega, wires=wires[0])]
+
+            # Otherwise recover the decomposition as a Rot, which can be further decomposed
+            # if desired. If the top left element is 0, can only use the off-diagonal elements
+            if qml.math.isclose(U[0, 0], 0):
+                phi = 1j * qml.math.log(U[0, 1] / U[1, 0])
+                omega = -phi - 2 * qml.math.angle(U[1, 0])
```

Comment:
You may need to be careful here:

- `phi` is complex
- `2 * angle(...)` will be real (I think `float64`)

and TensorFlow (maybe also PyTorch?) will not automatically cast the latter to a complex number before doing the subtraction.

I think the following should work:

```suggestion
                omega = -phi - qml.math.cast_like(2 * qml.math.angle(U[1, 0]), phi)
```

Reply:
(Note: `qml.math.cast(..., "complex128")` would also work above, but `cast_like` is safer, in that it will support both `complex64` and `complex128` depending on whatever `omega` is)

Reply:
Oh thanks, yes this looks safer! I originally had the conversion to `real` in the line of `phi` rather than in the return value, do you think it'd be better to convert `phi` first so that the subtract ends up real? Or is it better to just do the casting anyways?

Reply:
I think you'd need the casting regardless, potentially you could have float64 subtract float32, which would also cause an error